### PR TITLE
Add docker test to check that apparmor profile can be loaded without errors

### PR DESCRIPTION
**Pull request by andrii-suse**
Related to fix https://github.com/os-autoinst/openQA/pull/2413

To be able test apparmor the test needs to start docker container in --privileged mode, thus tests may impact host. (Be careful!)
In particular, 01-apparmor.t will reload apparmor profiles on host.
Docker service must be started, use one of commands below to run test: 
```
PRIVILEGED_TESTS=1 prove -v t/docker/01-apparmor.t
cd t/docker && PRIVILEGED_TESTS=1 ./01-apparmor.t
PRIVILEGED_TESTS=1 t/docker/01-apparmor.t
```

In circleci `--privileged` flag can be used only in 'machine' executor, so unfortunately `circleci local run` of the job is tricky to achieve.

### CODE REVIEW

File: script/osht.sh [hunk 1]
```diff
@@ -0,0 +1,470 @@
+###############################################################################
```
Comment (Martchus) [inline script/osht.sh]: This is a 3rd party file. When bundling dependencies it makes sense to provide a link to the original source and some versioning information. Since the abbreviation "osht" is meaningless to me it would also be nice to have a short description what this 3rd party script actually does.

File: script/test-in-container.sh [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+done
+
+docker exec "$containername" pwd >& /dev/null || (echo Cannot start container; exit 1 ) >&2
+
+if [ -n "$CIRCLECI" ]; then # circleci cannot mount volumes, so just copy current dir
+    docker cp "$thisdir"/../. "$containername:/opt/testing_area/"
+fi
+
+set +e
+docker exec -i "$containername" bash < "$testcase"
```
Comment (Martchus) [inline script/test-in-container.sh]: Why can't we just execute the testcase with `docker run` directly? These `docket exec` commands and the "sleep loop" seem needlessly complicating the test setup.

File: t/01-compile-check-all.t [hunk 3]
```diff
@@ -20,12 +20,24 @@ use FindBin;
...
 use Test::Strict;
+use Mojo::File;
 
 $Test::Strict::TEST_SYNTAX   = 1;
 $Test::Strict::TEST_STRICT   = 1;
 $Test::Strict::TEST_WARNINGS = 1;
-$Test::Strict::TEST_SKIP     = [
-    # skip test module which would require test API from os-autoinst to be present
-    't/data/openqa/share/tests/opensuse/tests/installation/installer_timezone.pm'
-];
```
Comment (Martchus) [inline t/01-compile-check-all.t]: Perl didn't like this?

```
$Test::Strict::TEST_SKIP     = [
    # skip test module which would require test API from os-autoinst to be present
    't/data/openqa/share/tests/opensuse/tests/installation/installer_timezone.pm',
    # problem: when tidy test fails - it leaves .tdy files, so we should ignore them
    Mojo::File->new->list_tree->grep(qr/\.tdy$/)->map('to_rel')->map('to_string')->to_array,
    # docker tests are in bash
    glob "t/docker/*.t",
];
```

Otherwise I'd prefer it over introducing 2 new variables and an additional push.

File: .circleci/config.yml [hunk 4]
```diff
@@ -271,6 +271,14 @@ jobs:
     <<: *test-template
     name: "Generating docs"
 
+  test.docker:
+    machine: true
+    steps:
+      - checkout
+      - run:
```
Comment (kalikiana) [inline .circleci/config.yml]: Please add `name: Run Docker tests` here

File: script/test-in-container.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+[[ ${BASH_SOURCE[0]} != *privileged.sh ]] || SKIP test "${PRIVILEGED_TESTS}" != 1 # PRIVILEGED_TESTS is not set to 1
+SKIP test "$(docker info 2>&1)" == "" # Docker doesn't seem to be running
+PLAN 1
+
+# let's use `docker build` here to untilize docker cache
+( 
+# shellcheck disable=SC2046 disable=SC2005
+echo "FROM registry.opensuse.org/devel/openqa/ci/containers/base:latest
+RUN sudo zypper ar -f -p 80 https://download.opensuse.org/repositories/devel:/openQA:/Leap:/15.1/openSUSE_Leap_15.1 openQA
+RUN sudo zypper ar -f -p 81 http://download.opensuse.org/repositories/devel:openQA/openSUSE_Leap_15.1 devel
```
Comment (okurz) [inline script/test-in-container.sh]: you have similar code covered already in multiple helper scripts. Please DRY. I can see you copy-pasted, because you used "https" for the first, "http" for the second in all cases ;)

File: script/test-in-container.sh [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+testcase=$1
+set -e pipefail
```
[the remark this answers is not in the document]
Comment (Martchus) [reply script/test-in-container.sh]: I don't prefer having this in the shebang because that's just a comment after all. If you invoke the script like `bash test-in-container.sh` the `-e` in that comment has no effect.

File: script/test-in-container.sh [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+thisdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+circlecidir="$thisdir"/../.circleci
+[ -d "$circlecidir" ] || circlecidir="$thisdir"/../../../.circleci
+[ -d "$circlecidir" ] || (echo Cannot find .circleci directory; exit 1 ) >&2
+basename=$(basename "$testcase")
+containername="localtest.$basename"
+
+# shellcheck source=/dev/null
+source "$thisdir/osht.sh"
```
[the remark this answers is not in the document]
Comment (Martchus) [reply script/test-in-container.sh]: Why? The `source` form is more readable in my opinion.

File: script/test-in-container.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (okurz) [reply script/test-in-container.sh]: no, the shebang is *not* a comment, weird convention maybe, yes, but it's not a comment. That you can override settings by calling it explicitly is a feature :)

File: script/test-in-container.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (okurz) [reply script/test-in-container.sh]: `.` is POSIX standard, `source` is non-standard: https://stackoverflow.com/questions/11588583/is-the-shells-source-posix-standard

File: script/test-in-container.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (Martchus) [reply script/test-in-container.sh]: From `bash`s perspective is *is* a comment. That the script behaves differently when invoking the interpreter manually is likely nothing one is usually expects so I consider it bad or maybe even a bug.

File: script/test-in-container.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (Martchus) [reply script/test-in-container.sh]: But this is a Bash script and not a generic shell script. I'm wondering why you don't like using Bash-features in Bash scripts.

File: script/test-in-container.sh [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+testcase=$1
+set -e pipefail
+
+[ -n "$testcase" ] || (echo No testcase provided; exit 1) >&2
+[ -f "$testcase" ] || (echo Cannot find file "$testcase"; exit 1 ) >&2
+
+thisdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```
[the remark this answers is not in the document]
Comment (perlpunk) [reply script/test-in-container.sh]: https://stackoverflow.com/a/35006505

File: script/test-in-container.sh [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+testcase=$1
+set -e pipefail
+
+[ -n "$testcase" ] || (echo No testcase provided; exit 1) >&2
```
[the remark this answers is not in the document]
Comment (coolo) [reply script/test-in-container.sh]: and *that* is POSIX?

[the remark this answers is not in the document]
Comment (okurz) [reply script/test-in-container.sh]: yes, it is

File: script/test-in-container.sh (hunk 2 above)
Comment (andrii-suse) [reply script/test-in-container.sh]: I believe this way it is easier to distinguish between 'test setup' and 'test run' (thus easier to troubleshoot eventual falures), plus it allows to mix commands in container and on host, the same way as it is done in `.circleci/build_dependencies.sh`

File: Makefile [hunk 10]
```diff
@@ -19,7 +19,7 @@ install:
 	done
 
 # we didn't actually want to install these...
-	for i in tidy check_coverage generate-packed-assets generate-documentation generate-documentation-genapi run-tests-within-container; do \
```
[the remark this answers is not in the document]
Comment (andrii-suse) [reply Makefile]: I see an error about `check_coverage is not a file` and cannot find it anywhere, so I believe that file was removed at some point?

File: script/test-in-container.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (andrii-suse) [reply script/test-in-container.sh]: I agree it is better, but next time changes whether file exists and it is better to have two similar checks than "one smart and the other"

File: script/test-in-container.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (andrii-suse) [reply script/test-in-container.sh]: yeah BASH_SOURCE is more reliable for edge case scenarios, otherwise it is the same - so why should we use `$0`?

File: script/test-in-container.sh (hunk 5 above)
Comment (andrii-suse) [reply script/test-in-container.sh]: DRY easily applies inside one system with defined or obvious layers, and introducing coupling between circleci scripts, docker containers and test scripts doesn't bring any real benefits.
At the moment I don't see any big advantage of replacing flat files with some hierarchy and e.g. sourcing scripts from inside heredoc which generates Dockerfile .

File: script/test-in-container.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (andrii-suse) [reply script/test-in-container.sh]: dirname returns relative path in some scenarios and having absolute path allows to not care whether current dir changed inside script

File: script/osht.sh (hunk 1 above)
Comment (andrii-suse) [reply script/osht.sh]: The test in circleci needs privileged container, thus must be run with 'machine' executor, which is Debian stretch. So I think the easiest way will be just downloading needed file from github at the moment.

File: script/test-in-container.sh (hunk 2 above)
Comment (Martchus) [reply script/test-in-container.sh]: I get your point but still don't think it is worth the benefit. But it isn't very important to me so keep it the way you like.

File: .circleci/dependencies.txt [hunk 11]
```diff
@@ -176,7 +176,7 @@ perl-Readonly-2.05
 perl-Regexp-Common-2017060201
 perl-Role-Tiny-2.000006
 perl-Scope-Guard-0.21
-perl-Selenium-Remote-Driver-1.34
+perl-Selenium-Remote-Driver-1.35
```
Comment (okurz) [inline .circleci/dependencies.txt]: I think these changes are already covered in another PR. I assume this diff will vanish after a rebase

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (okurz) [reply Makefile]: Please make that a separate commit

File: script/test-in-container.sh (hunk 2 above)
Comment (okurz) [reply script/test-in-container.sh]: I agree with @Martchus here, I would also prefer with a synchronous `docker run`

File: t/01-compile-check-all.t [hunk 12]
```diff
@@ -20,12 +20,19 @@ use FindBin;
...
 $Test::Strict::TEST_SYNTAX   = 1;
 $Test::Strict::TEST_STRICT   = 1;
 $Test::Strict::TEST_WARNINGS = 1;
-$Test::Strict::TEST_SKIP     = [
+
+$Test::Strict::TEST_SKIP = [
     # skip test module which would require test API from os-autoinst to be present
-    't/data/openqa/share/tests/opensuse/tests/installation/installer_timezone.pm'
+    't/data/openqa/share/tests/opensuse/tests/installation/installer_timezone.pm',
+    # problem: when tidy test fails - it leaves .tdy files, so we should ignore them
```
Comment (okurz) [inline t/01-compile-check-all.t]: I think you fixed a problem I never understood but have seen in some test runs, great!

File: script/test-in-container.sh (hunk 2 above)
Comment (andrii-suse) [reply script/test-in-container.sh]: One additional problem with `docker run bash -c "script"` is that it brings a challenge to pass variables and escape special characters properly.
And one more problem is that it is impossible to return into the container once `docker run` is finished, while splitting `run` and `exec` will give a lot of freedom into how container is used.
Last note is that I am working to Docker test which allows using systemd commands and it requires `docker run $image /usr/lib/systemd/systemd` and once the container is up - it is possible to `docker exec $container systemctl start myservice`, which will open path to test systemd scripts and more advanced scenarios locally.

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (andrii-suse) [reply Makefile]: @okurz if you think it needs separate commit - feel free to create another ticket and PR and I will use your commit here

File: script/test-in-container.sh (hunk 2 above)
Comment (okurz) [reply script/test-in-container.sh]: sounds great!

File: script/osht.sh (hunk 1 above)
Comment (okurz) [reply script/osht.sh]: yep, you do that already and that is fine.

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (okurz) [reply Makefile]: as discussed internally, sorry if I am annoying you. This was not my intention. For sure we should not need a separate ticket. https://github.com/os-autoinst/openQA/pull/2448 created for the code change in a separate commit.
